### ISSUE TITLE
CMake tip - if dispatcher generated cpp files are not rebuilding after changing a header

### ISSUE THREAD

**Issue by JoshBlake** [p12tic/libsimdpp#39]
(OSX 10.10, Apple Clang 7.0.2, CMake 3.5.0)

This is probably a bug in CMake (still looking into it) but I found it while working with libsimdpp, so thought other users might find this helpful.

## File setup

- src/main.cpp   //main program entry point
- src/code.cpp   //the code that dispatcher will copy
- src/common.h   //some common header, included in code.cpp and main.cpp
- include/simdpp/...  //simdpp include dir

CMakeLists.txt:
```
[...]
simdpp_get_runnable_archs(RUNNABLE_ARCHS)
simdpp_multiarch(GEN_ARCH_FILES src/code.cpp ${RUNNABLE_ARCHS})
add_executable(simd-test src/main.cpp ${GEN_ARCH_FILES})
target_include_directories(simd-test PRIVATE ${CMAKE_SOURCE_DIR}/include/)
```

## Background
The simdpp_multiarch() CMake function (from SimdppMultiarch.cmake) will use configure_file() to copy ${CMAKE_SOURCE_DIR}/src/code.cpp into the build dir (e.g. ${CMAKE_BINARY_DIR}/src/code_simdpp_-x86_avx.cpp etc). It will also manually add an the include dir back to the original location:
```
SimdppMultiarch.cmake line 434:
set(CXX_FLAGS "-I\"${CMAKE_CURRENT_SOURCE_DIR}/${SRC_PATH}\" ${CXX_FLAGS}")
```
This ensures that local includes, such as #include "common.h" in code.cpp will still work at compile time.

## Problem
The problem is that when CMake generates the file dependencies, it seems to ignore the file-specific include search path set on the generated files. This means that ${CMAKE_BINARY_DIR}/CMakeFiles/simd-test.dir/depend.make will _not_ include src/common.h and when you change common.h without changing code.cpp, none of the generated files are recompiled! This results in linking with stale object files (which include the old version of common.h) and programs that could crash or be incorrect in subtle ways.

## Workaround
Add the local directory of code.cpp to the target include dir with a command like this:
```
target_include_directories(simd-test PRIVATE ${PROJECT_SOURCE_DIR}/src)
```
(You may need to update the path for your project, or have multiple of these lines if you simdpp_multiarch() files from multiple directories.)

There may be a way to update simdpp_multiarch() to handle this automatically but a simple solution eludes me at the moment.

**Comment by JoshBlake** [comment]
Collaborating with @gordonmcshane, found an improvement to simdpp_multiarch() that avoids the need for a work-around. Will PR soon.

**Comment by p12tic** [comment]
@JoshBlake Does your fix involve `add_custom_command` and `IMPLICIT_DEPENDS` by chance?

**Comment by gordonmcshane** [comment]
@p12tic our fix was a bit more brute force.

In SimdppMultiarch.cmake, we just add the current source path to the current directory's INCLUDE_DIRECTORIES property via include_directories(). This works because INCLUDE_DIRECTORIES is inherited by subdirectories, and the generated code is placed in a child directory of the current source directory.

```
    # add current path to include paths so generated simd variations are
    # able to reference the same includes found in base version
    include_directories("${CMAKE_CURRENT_SOURCE_DIR}/${SRC_PATH}")

    foreach(ARCH ${ARCHS})
        simdpp_get_arch_info(CXX_FLAGS DEFINES_LIST SUFFIX ${ARCH})

        if(NOT "${SUFFIX}" STREQUAL "")
            # Copy the source file and add the required flags
            set(DST_ABS_FILE "${CMAKE_CURRENT_BINARY_DIR}/${SRC_PATH}/${SRC_NAME}_simdpp_${SUFFIX}${SRC_EXT}")
            set(SRC_ABS_FILE "${CMAKE_CURRENT_SOURCE_DIR}/${SRC_FILE}")
            configure_file("${SRC_ABS_FILE}" "${DST_ABS_FILE}" COPYONLY)
            list(APPEND FILE_LIST "${DST_ABS_FILE}")

            # let CMake know that this file is created at build time
            set_source_files_properties("${DST_ABS_FILE}" PROPERTIES GENERATED TRUE)

            if (NOT "${CXX_FLAGS}" STREQUAL "")
              set_source_files_properties("${DST_ABS_FILE}" PROPERTIES COMPILE_FLAGS ${CXX_FLAGS})
            endif()

            # For the first file that is being processed, set it to emit
            # dispatcher code. The required flags will be added later
            if("${DISPATCHER_FILE}" STREQUAL "")
                set(DISPATCHER_FILE "${DST_ABS_FILE}")
            endif()

            # Add required dispatcher predefined macros for this architecture
            set(DISPATCHER_CXX_FLAGS "${DISPATCHER_CXX_FLAGS} -DSIMDPP_DISPATCH_ARCH${DISPATCH_ARCH_IDX}=${DEFINES_LIST}")
            math(EXPR DISPATCH_ARCH_IDX "${DISPATCH_ARCH_IDX}+1")
        endif()
    endforeach()
```

Beyond the `include_directories()` statement, the only other change is the setting of `GENERATED` to `TRUE` on `DST_ABS_FILE` in the `foreach()`

**Comment by p12tic** [comment]
This could lead to issues, because `include_directories()` does not modify the include directories of source files. It sets the include directory for all targets defined in the current CMakeLists file and in CMakeLists files in children directories. These targets would end up with lots of extra include paths, which could cause problems if headers with same names are present.

I've committed the add_custom_command and IMPLICIT_DEPENDS fix at db8f68b2db06b050af.
